(PE-35138) Timeouts for task execution in bolt server

donoghuc opened the pull request:
TODO: Discuss on PR comments and decide what direction we want to go with timing out tasks run over ssh/winrm in PE

donoghuc commented:
We have three levels of "correctness" for stopping a running task over ssh/winrm (in rank order of correctness)
1. Stop the running task on the target host
2. Make bolt-server/ACE abandon the running task, but leave it running on the target
3. Make orchestrator abandon the running task, but leave bolt-server/ACE tracking it

I was hoping that closing the ssh session would cause any work to be done (at least over ssh) to terminate on the target. To test this I tired the following:
```diff
diff --git a/lib/bolt/transport/simple.rb b/lib/bolt/transport/simple.rb
index a04930b5..75c497fb 100644
--- a/lib/bolt/transport/simple.rb
+++ b/lib/bolt/transport/simple.rb
@@ -46,6 +46,10 @@ module Bolt

       def run_task(target, task, arguments, options = {}, position = [])
         with_connection(target) do |conn|
+          Thread.new do
+            sleep(10)
+            conn&.disconnect
+          end
           conn.shell.run_task(task, arguments, options, position)
         end
       end
```
I ran the following task against an ssh target
```ruby
#!/opt/puppetlabs/puppet/bin/ruby

sleep(100)
````
When I run this task via bolt CLI (with the hard coded timeout) I get:
```
(base) ➜  my_project bundle exec bolt task run my_project::foo -t rose-consultant.delivery.puppetlabs.net
Started on rose-consultant.delivery.puppetlabs.net...
forcing connection closed
#<Thread:0x00007f9cfc3e57b0@/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:471 run> terminated with exception (report_on_exception is true):
/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:109:in `select': Bad file descriptor (Errno::EBADF)
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:109:in `ev_select_and_postprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:30:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:227:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `block in loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/channel.rb:272:in `wait'
	from /Users/cas.donoghue/puppetlabs/bolt/lib/bolt/transport/ssh/connection.rb:218:in `block (2 levels) in execute'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
#<Thread:0x00007f9d0fa0ca68@/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:471 run> terminated with exception (report_on_exception is true):
/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:109:in `select': Bad file descriptor (Errno::EBADF)
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:109:in `ev_select_and_postprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:30:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:227:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `block in loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/channel.rb:272:in `wait'
	from /Users/cas.donoghue/puppetlabs/bolt/lib/bolt/transport/ssh/connection.rb:218:in `block (2 levels) in execute'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
#<Thread:0x00007f9cdc1c27c8@/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:471 run> terminated with exception (report_on_exception is true):
/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:75:in `select': closed stream (IOError)
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:75:in `available_for_read?'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:90:in `next_packet'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:193:in `block in poll_message'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:190:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:190:in `poll_message'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:545:in `dispatch_incoming_packets'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:248:in `ev_preprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:100:in `each'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:100:in `ev_preprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:28:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:227:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `block in loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/puppetlabs/bolt/lib/bolt/transport/ssh/connection.rb:202:in `block (2 levels) in execute'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
#<Thread:0x00007f9cdc1b9920@/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:471 run> terminated with exception (report_on_exception is true):
/Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:75:in `select': closed stream (IOError)
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:75:in `available_for_read?'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/packet_stream.rb:90:in `next_packet'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:193:in `block in poll_message'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:190:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/transport/session.rb:190:in `poll_message'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:545:in `dispatch_incoming_packets'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:248:in `ev_preprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:100:in `each'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:100:in `ev_preprocess'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/event_loop.rb:28:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:227:in `process'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `block in loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/net-ssh-6.1.0/lib/net/ssh/connection/session.rb:180:in `loop'
	from /Users/cas.donoghue/puppetlabs/bolt/lib/bolt/transport/ssh/connection.rb:202:in `block (2 levels) in execute'
	from /Users/cas.donoghue/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
Failed on rose-consultant.delivery.puppetlabs.net:
  closed stream
Failed on 1 target: rose-consultant.delivery.puppetlabs.net
Ran on 1 target in 15.92 sec
```
Unfortunately the task persists on the target:
```
[root@rose-consultant ~]# ps aux | grep ruby
root     19395  0.6  0.1 204996 14352 ?        Ss   20:59   0:00 /opt/puppetlabs/puppet/bin/ruby /tmp/b06a3eda-d2ae-4c0f-bf0d-ce1b5eb20bc7/foo.rb
```

After poking a bit more at that I decided to explore option `2`. I came up with starting the task execute in a future, then if we timeout waiting for that future raise an exepction (this could also be a new response). While this will result in the task being continued on the target, i *think* it should cause bolt-server to abandon it (I think that `wait_or_cancel` will handle cleaning up any of the outstanding work on the future.

mcdonaldseanp commented:
I don't think we can get away with anything other than the full "complete" solution of stopping the running task on the host. Otherwise we introduce inconsistencies in the way Orchestrator handles timeouts. The new approaches to task concurrency and scale that we've been talking about rely on consistent behavior across all types of task runs with respect to timeouts and stoppages. If we can ensure that all task PIDs running on all targets for a given task will be stopped after the specified timeout, there's a bunch of work that we can stop doing that Orchestrator does today to ensure it always has the status of every task.

Technically alot of that refactoring will be on the PCP side, but we still want to make sure that the new Architecture can rely on the backup of a timeout regardless of the transport.

We're also going to eventually need to allow for ad-hoc stoppages from users and from plans.

donoghuc commented:
I'm not sure I'm following the reasoning for that requirement. For ssh/winrm once bolt-server abandons the task that is it from orchestrator's perspective (we will record the task as a failure). For PCP where an agent is capable of reporting a success/failure that is a different story. Specifically it could be the case that we timeout but fail to kill the task and orchestrator could report that the task failed, but then in a world where we could then again ask the agent about the status of task it could be different if the task actually eventually finishes. 

I think that given ssh/winrm targets will never report the status/result from an abandon task it will not introduce any problems into our new model.

donoghuc commented:
Installed this and https://github.com/puppetlabs/orchestrator/pull/2142/files into a PE instance to manually poke around at this. Appears to be a viable path forward. Will plan on cleaning up the PRs after the long weekend :)

donoghuc commented on lib/bolt_server/transport_app.rb:
```diff
@@ -135,11 +136,30 @@ def task_helper(target, task, parameters)
+          end
+        end
+        task_future.execute
+        task_complete = task_future.wait_or_cancel(timeout)
```
The `wait_or_cancel` method does not actually do what I thought. The cancel only happens if the future does not actually begin, in the case of a timeout the future is not properly resolved. This is showing up in the rspec tests hanging in CI. I think I will need to use something else for this:
https://github.com/ruby-concurrency/concurrent-ruby/blob/master/docs-source/cancellation.in.md (see `Run only for limited time – Timeout replacement` section).

donoghuc replied:
UPDATE: given we are pinned back to an old concurrent ruby version I'm just going to do this with simple build in ruby threads.

nicklewis commented:
I'm a bit concerned about this. It looks correct to the extent that we're trying to implement a feature for _bolt-server_ (and insofar as `Thread#kill` is safe, which who knows), but I worry that it's effectively an endrun around the lack of timeout support in Bolt proper and that it will cause us pain in the future. If the answer is to kill a whole thread, we may be asking the wrong question...

Does this actually cause the running task to halt? If so, I am confused as to how (and curious whether we could emulate it). If not, then I'm not sure it's better than the initial implementation which was more to the point but only had the drawback of leaving the task running.

nicklewis commented:
Okay after reading the orchestrator PR, I do understand the logic here a bit better. This isn't a timeout for the user's sake, it's a timeout for orchestrator/bolt-server's sake, so it really _is_ an endrun around the lack of timeout support in Bolt. I can live with that, if we document it, though my question about what happens to running task and whether this implementation is better than the initial version still remains.

nicklewis commented on lib/bolt_server/transport_app.rb:
```diff
@@ -135,11 +135,30 @@ def task_helper(target, task, parameters)
-        value['_sensitive'] = value['_sensitive'].unwrap
+      if timeout
+        task_thread = Thread.new do
+          @executor.run_task(target, task, parameters).each do |result|
```
This copy-paste needs to be cleaned up. We also need to account for the case where the timeout is `0`.

donoghuc replied:
Oh, right I forgot about the special 0 case. Updated.

nicklewis commented on lib/bolt_server/transport_app.rb:
```diff
@@ -135,11 +135,30 @@ def task_helper(target, task, parameters)
+        if task_thread.join(timeout).nil?
+          task_thread.kill
+          raise Bolt::Error.new("Task execution on #{target.first.safe_name} timed out after #{timeout} seconds",
+                                'boltserver/task-timeout')
```
```suggestion
                                'bolt-server/task-timeout')
```

donoghuc replied:
I copied from https://github.com/puppetlabs/bolt/blob/a45eb27a775da0d9f0d2216c7d2c5477403bef84/lib/bolt_server/transport_app.rb#L874 and https://github.com/puppetlabs/bolt/blob/a45eb27a775da0d9f0d2216c7d2c5477403bef84/lib/bolt_server/transport_app.rb#L853

donoghuc commented:
Addressing your comments @nicklewis : 
>I'm a bit concerned about this. It looks correct to the extent that we're trying to implement a feature for bolt-server (and insofar as Thread#kill is safe, which who knows), but I worry that it's effectively an endrun around the lack of timeout support in Bolt proper and that it will cause us pain in the future. If the answer is to kill a whole thread, we may be asking the wrong question...

Is the introduction of the native ruby threads the root of the issue for you? I thought with the information I had provided above along with the info outlined in [my summary doc](https://perforcesoftwareinc-my.sharepoint.com/:w:/r/personal/cas_donoghue_perforce_com/Documents/Document1.docx?d=w57cd4cc90be249e7a027f4f3400c8f8f&csf=1&web=1&e=vethbh) we agreed that the problem we are solving immediately is ensuring orchestrator wont get stuck waiting for a hung ssh/winrm task and that for now we will focus on PCP and loop back to ssh/winrm later wrt actually killing execution. It is easy to abandon the task now in bolt-server to help us quickly move forward and it will require a more significant effort to ensure task execution stops on the target (a feature that may actually be useful/desirable in bolt). 

> Does this actually cause the running task to halt? If so, I am confused as to how (and curious whether we could emulate it). If not, then I'm not sure it's better than the initial implementation which was more to the point but only had the drawback of leaving the task running.

No, as far as I can see in testing `Thread.kill` is effectively the same as if you interrupt a bolt process after it starts the task. The task will continue to run on the target. This is consistend with the experiments I had done before which are documented in the comments here and the linked document. If you would like to poke around with this in a pe build I have it set up (along with corresponding orchestrator changes) on `slow-labyrinth.delivery.puppetlabs.net` (`vile-misfortune.delivery.puppetlabs.net` is an ssh inventory node).

mcdonaldseanp commented on lib/bolt_server/schemas/action-run_task.json:
```diff
@@ -9,7 +9,11 @@
-    "target": { "$ref": "partial:target-any" }
+    "target": { "$ref": "partial:target-any" },
+    "timeout": {
+      "type": "integer",
```
Does anything check that the integer is positive?

mcdonaldseanp replied:
wait, ignore me, we absolutely do with timeout > 0

mcdonaldseanp reviewed (approved):
LGTM, just one question about checking for positive integers. Not sure if 1. we care that much or 2. if something already is checking for it to be positive, so I approve :D
